core/src/main/java/org/elasticsearch/indices/query/IndicesQueriesRegistry.java

```diff
@@ -19,24 +19,47 @@
 
 package org.elasticsearch.indices.query;
 
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.ParseFieldMatcher;
+import org.elasticsearch.common.ParsingException;
 import org.elasticsearch.common.component.AbstractComponent;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentLocation;
 import org.elasticsearch.index.query.QueryParser;
 
 import java.util.Map;
 
 public class IndicesQueriesRegistry extends AbstractComponent {
     private Map<String, QueryParser<?>> queryParsers;
+    private Map<String, ParseField> parseFields;
+    private ParseFieldMatcher parseFieldMatcher;
 
-    public IndicesQueriesRegistry(Settings settings, Map<String, QueryParser<?>> queryParsers) {
+    public IndicesQueriesRegistry(Settings settings, Map<String, QueryParser<?>> queryParsers, Map<String, ParseField> parseFields) {
         super(settings);
+        this.parseFieldMatcher = new ParseFieldMatcher(settings);
         this.queryParsers = queryParsers;
+        this.parseFields = parseFields;
+        //TODO validate size of the two maps
     }
 
     /**
-     * Returns all the registered query parsers
+     * Get the query parser for a specific type of query registered under its name.
+     * Uses {@link ParseField} internally so that deprecation warnings/errors can be logged/thrown.
+     * @param name the name of the parser to retrieve
+     * @return the query parser
+     * @throws IllegalArgumentException of there's no query or parser registered under the provided name
      */
-    public Map<String, QueryParser<?>> queryParsers() {
-        return queryParsers;
+    public QueryParser<?> getQueryParser(String name, XContentLocation xContentLocation) {
+        ParseField parseField = parseFields.get(name);
+        if (parseField == null) {
+            throw new ParsingException(xContentLocation, "No query registered for [" + name + "]");
+        }
+        boolean match = parseFieldMatcher.match(name, parseField);
+        assert match : "registered parse field did not match the query name it was registered for";
+        QueryParser<?> queryParser = queryParsers.get(name);
+        if (queryParser == null) {
+            throw new ParsingException(xContentLocation, "No parser registered for [" + name + "]");
```

Comment:
Should this also be an assert since it should not be possible to register a parseField without a queryParser?